File: grasshopper_mcp_server/src/grasshoppermcp/tools/bake_component.py
"""Bake a component's geometry to the Rhino document."""

from mcp.server.fastmcp import Context
from grasshoppermcp.server import get_grasshopper_connection, mcp, logger
from typing import Dict, Any, Optional, List


@mcp.tool()
def bake_component(
    ctx: Context,
    instance_id: Optional[str] = None,
    nickname: Optional[str] = None,
    output_index: int = 0,
    layer_name: Optional[str] = None
) -> Dict[str, Any]:
    """
    Bake a component's geometry output to the Rhino document.

    Parameters:
    - instance_id: GUID of the component to bake
    - nickname: Nickname of the component to bake (alternative to instance_id)
    - output_index: Index of the output parameter to bake (default: 0)
    - layer_name: Name of the layer to bake to (creates if doesn't exist)

    At least one of instance_id or nickname must be provided.

    Returns:
    - baked_count: Number of objects baked
    - object_ids: List of GUIDs of the baked Rhino objects
    - layer: Name of the layer objects were baked to
    - message: Confirmation message

    Example:
        # Bake a circle to Rhino
        result = bake_component(nickname="MyCircle")
        print(f"Baked {result['baked_count']} objects")

        # Bake to a specific layer
        bake_component(nickname="MyExtrusion", layer_name="Grasshopper Bake")
    """
    try:
        if not instance_id and not nickname:
            return {"success": False, "message": "Either instance_id or nickname is required"}

        gh = get_grasshopper_connection()
        params = {"output_index": output_index}
        if instance_id:
            params["instance_id"] = instance_id
        if nickname:
            params["nickname"] = nickname
        if layer_name:
            params["layer_name"] = layer_name

        result = gh.send_command("bake_component", params)
        return {"success": True, **result}
    except Exception as e:
        logger.error(f"Error baking component: {str(e)}")
        return {"success": False, "message": str(e)}

File: grasshopper_mcp_plugin/Functions/BakeComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Grasshopper;
using Grasshopper.Kernel;
using Newtonsoft.Json.Linq;
using Rhino;
using Rhino.DocObjects;

namespace GrasshopperMCPPlugin.Functions;

public partial class GrasshopperMCPFunctions
{
    /// <summary>
    /// Bake geometry from a component to the Rhino document.
    /// Supports both Python naming (instance_id, nickname, output_index)
    /// and C# naming (component_id, param_index).
    /// </summary>
    public JObject BakeComponent(JObject parameters)
    {
        var ghDoc = Instances.ActiveCanvas?.Document;
        var rhinoDoc = RhinoDoc.ActiveDoc;

        if (ghDoc == null)
        {
            throw new InvalidOperationException("No active Grasshopper document");
        }

        if (rhinoDoc == null)
        {
            throw new InvalidOperationException("No active Rhino document");
        }

        // Find component - supports instance_id, nickname, or component_id
        var obj = ComponentHelper.FindComponent(ghDoc, parameters);

        // Get layer name (support both layer_name and layer)
        var layerName = parameters["layer_name"]?.ToString()
                     ?? parameters["layer"]?.ToString();

        // Get output parameter
        var paramIndex = ComponentHelper.GetParamIndex(parameters, isOutput: true);
        var paramName = ComponentHelper.GetParamName(parameters, isOutput: true);

        var outputParam = ComponentHelper.FindOutputParam(obj, paramIndex, paramName);

        // If no specific output found, try to find first geometry output
        if (outputParam == null && obj is IGH_Component component)
        {
            outputParam = component.Params.Output.FirstOrDefault(p =>
                p.TypeName.Contains("Geometry") ||
                p.TypeName.Contains("Curve") ||
                p.TypeName.Contains("Surface") ||
                p.TypeName.Contains("Brep") ||
                p.TypeName.Contains("Mesh") ||
                p.TypeName.Contains("Point"));

            if (outputParam == null && component.Params.Output.Count > 0)
                outputParam = component.Params.Output[0];
        }

        if (outputParam == null)
        {
            throw new InvalidOperationException("Could not find output parameter to bake");
        }

        // Check if parameter supports baking
        if (!(outputParam is IGH_BakeAwareObject bakeAware))
        {
            throw new InvalidOperationException($"Parameter '{outputParam.Name}' does not support baking");
        }

        // Set up bake attributes
        var attributes = new ObjectAttributes();

        // Set layer if specified
        if (!string.IsNullOrEmpty(layerName))
        {
            var layerIndex = rhinoDoc.Layers.FindByFullPath(layerName, -1);
            if (layerIndex < 0)
            {
                // Create layer if it doesn't exist
                var layer = new Layer { Name = layerName };
                layerIndex = rhinoDoc.Layers.Add(layer);
            }
            attributes.LayerIndex = layerIndex;
        }

        // Bake the geometry
        var bakedIds = new List<Guid>();

        bakeAware.BakeGeometry(rhinoDoc, attributes, bakedIds);

        rhinoDoc.Views.Redraw();

        // Format result
        var resultIds = new JArray();
        foreach (var id in bakedIds)
        {
            resultIds.Add(id.ToString());
        }

        return new JObject
        {
            ["instance_id"] = obj.InstanceGuid.ToString(),
            ["nickname"] = obj.NickName,
            ["param_name"] = outputParam.Name,
            ["baked_count"] = bakedIds.Count,
            ["object_ids"] = resultIds,
            ["layer"] = layerName ?? "Default",
            ["message"] = $"Baked {bakedIds.Count} object(s) from {obj.NickName}.{outputParam.Name}"
        };
    }
}
